src/idl_parser.cpp

```diff
@@ -3266,19 +3279,40 @@ CheckedError Parser::ParseRoot(const char *source, const char **include_paths,
       for (auto val_it = enum_def.Vals().begin();
            val_it != enum_def.Vals().end(); ++val_it) {
         auto &val = **val_it;
+        if (val.union_type.struct_def) {
+          auto err = CheckPrivateLeak(enum_def,
+                                      *val.union_type.struct_def);
+          if (err.Check()) { return err; }
+        }
+
         if (!(opts.lang_to_generate != 0 && SupportsAdvancedUnionFeatures()) &&
             (IsStruct(val.union_type) || IsString(val.union_type)))
+
           return Error(
               "only tables can be union elements in the generated language: " +
               val.name);
       }
     }
   }
+
   // Parse JSON object only if the scheme has been parsed.
   if (token_ == '{') { ECHECK(DoParseJson()); }
   return NoError();
 }
 
+CheckedError Parser::CheckPrivateLeak(const Definition &def,
+                                      const Definition &value_type) {
+  if (!opts.no_leak_private_annotations) return NoError();
```

Comment:
Do we want to warn users if they're leaking private annotations?

I think we need to disallow leaking in the next major release and adding a warning will help transition people. But then again, new warning are breaking changes if warnings are errors.

This could be considered a future problem so no action requested

Reply:
Yeah, they are, I'm not sure if we have a warning flag though. But it seems that `error_` will only be emitted to the user when the parsing fails.